File: Data Structures & Algorithms/matrixDFS/submission-0.py
class Solution:
    def countPaths(self, grid: List[List[int]]) -> int:
        visit = set()
        ROWS = len(grid)
        COLS = len(grid[0])
        def dfs(r,c):
            if r<0 or r>= ROWS or c<0 or c>=COLS or (r,c) in visit or grid[r][c] ==1:
                return 0
            if r == ROWS -1 and c== COLS-1:
                return 1
            
            visit.add((r,c))
            res = dfs(r+1,c) + dfs(r,c+1) + dfs(r-1,c) + dfs(r,c-1)
            visit.remove((r,c))
            return res
        return dfs(0,0)
            
    

File: Data Structures & Algorithms/count-number-of-islands/submission-0.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int res = 0;
        for (int i = 0; i < grid.size(); i++ ){
            for (int j = 0; j < grid[0].size(); j++){
                if (grid[i][j] == '1'){
                    dfs(i, j, grid);
                    res++;
                }

            }
        }
        return res;
        
    }
    void dfs(int r, int c, vector<vector<char>>& grid){
        if (r < 0 || r >= grid.size() || c < 0 || c >= grid[0].size() || grid[r][c] == '0'){
            return;
        }
        grid[r][c] = '0';

        dfs(r, c+1, grid);
        dfs(r, c-1, grid);
        dfs(r-1,c,grid);
        dfs(r+1,c, grid);

    }
};